File: xllm/core/layers/dcu/attention.cpp
/* Copyright 2025-2026 The xLLM Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    https://github.com/jd-opensource/xllm/blob/main/LICENSE

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
==============================================================================*/

#include "layers/dcu/attention.h"

#include <cstdlib>
#include <cstring>

#include "core/util/rec_model_utils.h"
#include "layers/dcu/base_attention_impl.h"
#include "layers/dcu/flash_attention.h"
#include "layers/dcu/torch_attention.h"

namespace xllm {
namespace layer {
AttentionImpl::AttentionImpl(int64_t num_heads,
                             int64_t head_size,
                             float scale,
                             int64_t num_kv_heads,
                             int64_t sliding_window) {
  // Select implementation based on mode. Use polymorphism via base class
  // pointer to manage different implementations.
  VLOG(1) << "## scale " << scale << " num_heads " << num_heads << " head_size "
          << head_size << " num_kv_heads " << num_kv_heads << " sliding_window "
          << sliding_window << " is_rec_multi_round_mode "
          << is_rec_multi_round_mode();

  // Debug switch: set XLLM_DCU_ATTN_IMPL=torch to fall back to the pure-PyTorch
  // implementation (TorchAttentionImpl) instead of the default FlashAttention
  // kernel. Useful for cross-checking numerics when debugging the fast path.
  const char* impl_env = std::getenv("XLLM_DCU_ATTN_IMPL");
  if (impl_env != nullptr && std::strcmp(impl_env, "torch") == 0) {
    LOG(INFO) << "AttentionImpl: using TorchAttentionImpl (debug fallback via "
                 "XLLM_DCU_ATTN_IMPL=torch)";
    attention_impl_ = std::make_shared<TorchAttentionImpl>(
        num_heads, head_size, scale, num_kv_heads, sliding_window);
  } else {
    attention_impl_ = std::make_shared<FlashAttentionImpl>(
        num_heads, head_size, scale, num_kv_heads, sliding_window);
  }
}

std::tuple<torch::Tensor, std::optional<torch::Tensor>> AttentionImpl::forward(
    const AttentionMetadata& attn_metadata,
    torch::Tensor& query,
    torch::Tensor& key,
    torch::Tensor& value,
    KVCache& kv_cache) {
  // Create output tensor internally to unify the interface with other devices
  torch::Tensor output = torch::empty_like(query);

  VLOG(1) << "AttentionImpl::forward [CUDA graph mode]:   "
          << attn_metadata.enable_cuda_graph;

  if (!attn_metadata.enable_cuda_graph) {
    VLOG(1)
        << "AttentionImpl::forward"
        << " is_prefill=" << attn_metadata.is_prefill
        << " batch_size=" << (attn_metadata.kv_cu_seq_lens.size(0) - 1)
        << " max_seq_len=" << attn_metadata.max_seq_len
        << " max_query_len=" << attn_metadata.max_query_len
        << " q_shape=" << query.sizes() << " k_shape=" << key.sizes()
        << " v_shape=" << value.sizes() << " output_shape=" << output.sizes()
        << " kv_cache k_cache=" << kv_cache.get_k_cache().sizes()
        << " kv_cache v_cache=" << kv_cache.get_v_cache().sizes()
        << " full_k_cache size =" << attn_metadata.full_k_cache.sizes()
        << " full_v_cache=" << attn_metadata.full_v_cache.sizes()
        << " unshared_k_cache size =" << attn_metadata.unshared_k_cache.sizes()
        << " unshared_v_cache size =" << attn_metadata.unshared_v_cache.sizes()
        << " q_cu_seq_lens=" << attn_metadata.q_cu_seq_lens
        << " kv_cu_seq_lens=" << attn_metadata.kv_cu_seq_lens
        << " block_table=" << attn_metadata.block_table.sizes()
        << " paged_kv_indptr size =" << attn_metadata.paged_kv_indptr.sizes()
        << " paged_kv_indices size =" << attn_metadata.paged_kv_indices.sizes()
        << " paged_kv_last_page_len=" << attn_metadata.paged_kv_last_page_len
        << " slot_mapping size =" << attn_metadata.slot_mapping.sizes()
        << " q_seq_lens=" << attn_metadata.q_seq_lens
        << " kv_seq_lens=" << attn_metadata.kv_seq_lens;
  } else {
    // CUDA graph capture mode: cannot call tensor methods (sizes(), defined(),
    // numel()) but can print scalar fields and tensor objects directly
    VLOG(1) << "AttentionImpl::forward [CUDA graph mode]"
            << " kv_cache k_cache=" << kv_cache.get_k_cache().sizes()
            << " kv_cache v_cache=" << kv_cache.get_v_cache().sizes()
            << " is_prefill=" << attn_metadata.is_prefill
            << " max_seq_len=" << attn_metadata.max_seq_len
            << " max_query_len=" << attn_metadata.max_query_len
            << " total_kv_len=" << attn_metadata.total_kv_len
            << " is_causal=" << attn_metadata.is_causal
            << " is_chunked_prefill=" << attn_metadata.is_chunked_prefill
            << " is_dummy=" << attn_metadata.is_dummy
            << " compute_dtype=" << attn_metadata.compute_dtype
            << " enable_cuda_graph=" << attn_metadata.enable_cuda_graph;
  }

  // Use polymorphism to dispatch to the appropriate implementation,
  // making the code elegant and type-safe.
  return attention_impl_->forward(
      attn_metadata, query, key, value, output, kv_cache);
}

}  // namespace layer
}  // namespace xllm

File: xllm/python/kernels_npu/attention.py
# Copyright 2026 The xLLM Authors.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://github.com/xLLM-AI/xllm/blob/main/LICENSE
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""NPU paged-attention support kernels.

Both operators are capturable: they read only their arguments, so a decode
graph can replay them without re-planning.
"""

from __future__ import annotations

import torch

reshape_paged_cache = torch.ops.xllm_ops.reshape_paged_cache
update_decode_graph_metadata = torch.ops.xllm_ops.update_decode_graph_metadata


def vision_fusion_attention(
    q: torch.Tensor,
    k: torch.Tensor,
    v: torch.Tensor,
    *,
    actual_seq_qlen: list[int],
    actual_seq_kvlen: list[int],
    num_heads: int,
    scale: float,
    input_layout: str = "TND",
) -> torch.Tensor:
    """Run fused self-attention for ViT blocks via ``torch_npu.npu_fusion_attention``.

    Args:
        q, k, v: Query/key/value in the layout specified by ``input_layout``.
        actual_seq_qlen: Cumulative query sequence lengths.
        actual_seq_kvlen: Cumulative key/value sequence lengths.
        num_heads: Number of attention heads.
        scale: Attention scaling factor.
        input_layout: Tensor layout (default ``"TND"``).

    Returns:
        Attention output with same shape as ``q``.
    """
    import torch_npu

    return torch_npu.npu_fusion_attention(
        q,
        k,
        v,
        actual_seq_qlen=actual_seq_qlen,
        actual_seq_kvlen=actual_seq_kvlen,
        head_num=num_heads,
        scale=scale,
        input_layout=input_layout,
    )[0]


__all__ = [
    "reshape_paged_cache",
    "update_decode_graph_metadata",
    "vision_fusion_attention",
]
